[Enhancement] Enhance AstToSQLBuilder formatting for CASE-WHEN and CTE

wangyang377 opened the pull request:
Improve SQL formatting capabilities in AstToSQLBuilder to provide better readability for complex SQL structures.

- Format each SELECT column on a separate line with proper indentation
- Format CASE-WHEN with proper indentation for WHEN/ELSE/END clauses
- Format CTE definitions with proper multi-line structure
- Format multiple CTEs with comma at line end
- Add indentation management methods to FormatOptions

Fixes: #64056

## Why I'm doing:

## What I'm doing:

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Formats SELECT items on separate indented lines, adds proper CASE/WHEN and CTE (WITH) multi-line formatting, and introduces indentation utilities with updated tests.
> 
> - **SQL formatting enhancements**:
>   - `SELECT` items rendered on separate lines with indentation.
>   - `CASE ... WHEN ... THEN ... ELSE ... END` formatted across lines with consistent indent.
>   - CTEs in `WITH` clause formatted as multi-line blocks; supports multiple CTEs separated by comma and newline.
>   - `visitQueryStatement` handles `WITH`, main query, `ORDER BY`, and `LIMIT` output sequencing.
> - **Indentation utilities**:
>   - Add `indentLevel`, `indentString`, `indent()`, `increaseIndent()`, `decreaseIndent()`, `getIndentLevel()` in `FormatOptions`.
> - **Tests**:
>   - Update expectations for multi-line `SELECT` outputs.
>   - Add tests for CASE formatting, CTE formatting, and complex nested CTE with CASE.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d2437f706807656e2fd72af3ea796187175592a7. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

wangyang377 committed:
[Enhancement] Enhance AstToSQLBuilder formatting for CASE-WHEN and CTE

Improve SQL formatting capabilities in AstToSQLBuilder to provide
better readability for complex SQL structures.

- Format each SELECT column on a separate line with proper indentation
- Format CASE-WHEN with proper indentation for WHEN/ELSE/END clauses
- Format CTE definitions with proper multi-line structure
- Format multiple CTEs with comma at line end
- Add indentation management methods to FormatOptions

Fixes: #64056
Signed-off-by: wangyang <wangyang377@ustc.edu>

alvin-celerdata commented:
@cursor review

wangyang377 committed:
[BugFix] Fix CTE closing parenthesis indentation in AST2SQLVisitor

The visitCTE method incorrectly formatted the closing parenthesis
of a CTE definition. After the CTE query, the indent level was
decreased, but the closing parenthesis was then appended on a new
line without applying the correct indentation, causing it to appear
at column 0 instead of aligning properly.

- Changed from options.newLine() + append(")") to options.indent().append(")")
- This ensures the closing parenthesis has proper indentation
- Aligns with the pattern used in visitCaseWhenExpr method

Signed-off-by: wangyang690 <wangyang690@jd.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StarRocks/starrocks?pullRequest=64237) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/StarRocks/starrocks?pullRequest=64237) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: wangyang377<br/>:x: wangyang690<br/><hr/>**wangyang690** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StarRocks/starrocks?pullRequest=64237) it.</sub>

wangyang377 commented on fe/fe-core/src/main/java/com/starrocks/sql/formatter/AST2SQLVisitor.java:
```diff
@@ -262,7 +280,68 @@ public String visitCTE(CTERelation relation, Void context) {
+        sqlBuilder.append(options.newLine());
+        sqlBuilder.append(")");
+        
+        return sqlBuilder.toString();
```
Changed to use options.indent() which includes both newline and proper indentation:sqlBuilder.append(options.indent()).append(")");

murphyatwork commented on fe/fe-core/src/test/java/com/starrocks/sql/analyzer/AstToSQLBuilderTest.java:
```diff
@@ -63,33 +63,127 @@ public void testInsertFromFiles() {
+    public void testCaseWhenFormatting() {
+        String sql = "SELECT CASE WHEN v1 < 10 THEN 'low' WHEN v1 >= 10 AND v1 < 20 THEN 'medium' ELSE 'high' END FROM t0";
+        StatementBase stmt = SqlParser.parseSingleStatement(sql, SqlModeHelper.MODE_DEFAULT);
+        String expected = "SELECT \n" +
```
other syntax uses the `XXX\n` rather than `XXX \n`, so for consistency we should remove the ` ` here

wangyang377 committed:
[Enhancement] Add enablePrettyFormat option to control SQL formatting

Add a new enablePrettyFormat option to FormatOptions to control
whether to use pretty formatting with multi-line indentation.
This option is disabled by default to maintain backward compatibility.

- Add enablePrettyFormat flag to FormatOptions (default: false)
- Update visitSelect, visitCTE, visitQueryStatement, visitCaseWhenExpr
  to conditionally apply pretty formatting based on the flag
- Add toPrettySQL helper method in AstToSQLBuilderTest
- Update test cases for CASE-WHEN and CTE formatting to use pretty format
- Keep existing tests unchanged (use default non-pretty format)

Benefits:
- Backward compatible: default behavior remains unchanged
- Opt-in feature: only tests that need pretty format enable it explicitly
- No impact on production code: all existing usages continue to work
- Clean separation: formatting logic is conditionally applied

Signed-off-by: wangyang690 <wangyang690@jd.com>

wangyang377 committed:
[Enhancement] Improve SQL formatting logic in AST2SQLVisitor

- Move visitSelectItemList call to respective format branches

- Add try-finally block for indent management in pretty format

- Fix CTE closing parenthesis spacing in default format

Signed-off-by: wangyang690 <wangyang690@jd.com>
